src/asl.ts

```diff
@@ -0,0 +1,720 @@
+              Next: this.throw(expr),
+            },
+          ],
+          Next: this.next(expr),
+        };
+      }
+      throw new Error(`call must be a service call, ${expr}`);
+    } else if (
+      expr.kind === "Identifier" ||
+      expr.kind === "PropAccessExpr" ||
+      expr.kind === "ElementAccessExpr" ||
+      isLiteralExpr(expr)
+    ) {
+      return {
+        Type: "Pass",
+        Next: this.next(expr),
+        Parameters: {
+          [`result${isLiteralExpr(expr) ? "" : ".$"}`]: this.evalJson(expr),
+        },
+        OutputPath: "$.result",
+        ...(ResultPath ? { ResultPath } : {}),
+      };
+    } else {
+      throw new Error(`cannot eval expression kind '${expr.kind}'`);
+    }
+  }
+
+  public evalJson(expr: Expr): any {
+    if (expr.kind === "Identifier") {
+      return this.evalJsonPath(expr);
+    } else if (expr.kind === "PropAccessExpr") {
+      return `${this.evalJson(expr.expr)}.${expr.name}`;
+    } else if (
+      expr.kind === "ElementAccessExpr" &&
+      expr.element.kind === "NumberLiteralExpr"
+    ) {
+      return `${this.evalJson(expr.expr)}[${expr.element.value}]`;
+    } else if (expr.kind === "ArrayLiteralExpr") {
+      return expr.items.map(this.evalJson);
+    } else if (expr.kind === "ObjectLiteralExpr") {
+      const payload: any = {};
+      for (const prop of expr.properties) {
+        if (prop.kind !== "PropAssignExpr") {
+          throw new Error(
+            `${prop.kind} is not supported in Amazon States Language`
+          );
+        }
+        if (
+          prop.name.kind !== "StringLiteralExpr" &&
+          prop.name.kind !== "Identifier"
+        ) {
+          throw new Error(
+            `computed name of PropAssignExpr is not supported in Amazon States Language`
+          );
+        }
+
+        payload[
+          `${
+            prop.name.kind === "StringLiteralExpr"
+              ? prop.name.value
+              : prop.name.name
+          }${
+            isLiteralExpr(prop.expr) || isReferenceExpr(prop.expr) ? "" : ".$"
+          }`
+        ] = this.evalJson(prop.expr);
+      }
+      return payload;
+    } else if (isLiteralExpr(expr)) {
+      return expr.value;
+    } else if (expr.kind === "ReferenceExpr") {
+      const ref = expr.ref();
+      if (ref.kind === "Function") {
+        return ref.resource.functionArn;
+      } else if (ref.kind === "StepFunction") {
+        return ref.stateMachineArn;
+      } else if (ref.kind === "Table") {
+        return ref.resource.tableName;
+      }
+    }
+    throw new Error(`cannot evaluate ${expr.kind} to JSON`);
+  }
+
+  public condition(expr: Expr): Condition {
+    if (expr.kind === "UnaryExpr") {
+      return {
+        Not: this.condition(expr.expr),
+      };
+    } else if (expr.kind === "BinaryExpr") {
+      if (expr.op === "&&") {
+        return {
+          And: [this.condition(expr.left), this.condition(expr.right)],
+        };
+      } else if (expr.op === "||") {
```

Comment:
Any plans to reduce nested ANDs?

Reply:
Is there a good reason to? For readability?

Reply:
Not sure, just less json output? Maybe better performance? But yeah, probably no real impact.